### PR TITLE
No remote bot attack  - Empty array shouldn't allow access, Test with un...

### PR DESCRIPTION
**Pull request by butchs**
...set($user_info) prior to change. -s

### CODE REVIEW

**Comment by emanuele45** [comment]
May I ask you the action the bot is trying when you get this error?
Asking because is_not_guest is called in few places and some of them already check for is_guest, so I'd like to have an idea what the bot is doing. ;)

**Comment by butchs** [comment]
I did not look at the other places.  I was merly testing the new code in my test server.  I stumbled into it when playing with the adminLogin password re-enter check via validate session.

Some nasty bots scan for errors and run patterns to gain access.  This function can be compromised.   I was able to gain access to my admin section even though $user_info was empty.  A bot can do this by running the function or series of functions remotely while the admin is logged in.

**Comment by emanuele45** [comment]
> This function can be compromised. I was able to gain access to my admin section even though $user_info was empty.

How? Details.

**Comment by mikemill** [comment]
@emanuele45 this one would be "more correct" than #184 as 184 is merging into master where this one is merging into the release-2.1 branch (aka the "correct" branch).

**Comment by emanuele45** [comment]
Yep, that's why this one is still open and #184 is closed.

**Comment by mikemill** [comment]
Bah, GitHub confused my sleepy state :(

**Comment by butchs** [comment]
Even if i knew, I know nothing of the sort.  

Playing with code is a  long long time hobby for me.  But alas, it does not pay the bills.  My other talents have become more profitable.  When I was between jobs a few years ago I wrote my Firewall program.  I wrote functions that properly set arrays in one file, then called a function in another file only to see undefined errors.  I was able to catch the bot to confirm my suspicion.

I tested and confirmed that some bots scan and run functions.  Errors were created even when data was properly set in one file and immediately sent to only one function another file.   I revised my code with validation methods to protect against such attempts.  Believe me or not!

At first I thought, it could be because SMF and I are not using access control modifiers.  But php is a scripted language and some view them as mere "purposeless access decorators".   Instead I looked at proactive access control methods that could stop someone with the free time and experience, not that i know anything about it, to attack the function that could assist to gain access.  My methods are not published...  Because they are unique.

Here is another link:
https://www.owasp.org/index.php/Access_Control_Cheat_Sheet
Read "Never Depend on Untrusted Data".   I add that the data structure of the input should never be trusted.  When it comes to Admin access the data should be validated.  Sorry but I do not trust set values.  Believe me or not!

I have to apologize I messed up the code a tad...  The code should be "if (!is_array($user_info) && !$user_info['is_guest'])"

There is no error log.  All I get is "Undefined index: is_guest" when I force the function.  Like I said I debug live.  I edit the code on my HD and test it live. Like an assembly debugger but for php.

I am only looking at the admin login because that is what can cause the most damage.  Here is what I did:

Located the function "is_not_guest".
Add "unset($user_info['is_guest']);" after the global line.
Viewing your site as a guest click on Login in "Welcome, Guest. Please login or register."
Enter the admin users and and password.
Then try to access the admin section.
Repeat with and without the change.

If you have a better way go ahead and do it. 

-s

**Comment by norv** [comment]
I would make this change. Precisely because there are multiple execution paths that need this, and is_not_guest() may also be used from external code.

File: Sources/Security.php [hunk 1]
```diff
@@ -105,8 +105,8 @@ function is_not_guest($message = '')
 {
 	global $user_info, $txt, $context, $scripturl;
 
-	// Luckily, this person isn't a guest.
-	if (!$user_info['is_guest'])
+	// Luckily, this person isn't bot and a guest.
+	if (is_array($user_info) && !$user_info['is_guest'])
```
Comment (butchs) [inline Sources/Security.php:109]: if (!is_array($user_info) && !$user_info['is_guest'])   //  correct the typo as pointed out in discussion

**Comment by emanuele45** [comment]
Fixed in 26234d3530ea3fe33c5fab5ff1fa4086faf31da8 with isset

**Pull request closed without merging**